### PR TITLE
Built a modal system for the UI

### PR DESCRIPTION
**Pull request by deammer**
We're gonna need some modals so we can nest forms within forms within forms 😭 

### Changes

- added a modal system
- added the `body-scroll-lock` package which prevents users from scrolling the page when a modal is visible
- added the `react-transition-group` package for animations

https://user-images.githubusercontent.com/3411183/111924645-d337a700-8a62-11eb-965f-70038986e29c.mov

### CODE REVIEW

File: frontend/src/components/animation/TailwindTransition.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import { FunctionComponent } from 'react'
+import { CSSTransition as ReactCSSTransition } from 'react-transition-group'
```
Comment (bion) [inline frontend/src/components/animation/TailwindTransition.tsx:2]: 🎶hello darkness my old friend🎶

File: frontend/src/components/animation/TailwindTransition.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  leaveFrom = '',
+  leaveTo = '',
+  ...otherProps
+}) => {
+  const enterClasses = enter.split(' ').filter((s) => s.length > 0)
+  const enterFromClasses = enterFrom.split(' ').filter((s) => s.length > 0)
+  const enterToClasses = enterTo.split(' ').filter((s) => s.length > 0)
+  const leaveClasses = leave.split(' ').filter((s) => s.length > 0)
+  const leaveFromClasses = leaveFrom.split(' ').filter((s) => s.length > 0)
+  const leaveToClasses = leaveTo.split(' ').filter((s) => s.length > 0)
```
Comment (bion) [inline frontend/src/components/animation/TailwindTransition.tsx]: suggestion (blocking): please extract a method for `.split(' ').filter((s) => s.length > 0)` with a name that describes why it's necessary or a comment if clear name for that isn't easy

File: frontend/src/components/animation/TailwindTransition.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    if (classes.length > 0) {
+      node.classList.add(...classes)
+    }
+  }
+
+  const removeClasses = (node: HTMLElement, classes: string[]) => {
+    if (classes.length > 0) {
+      node.classList.remove(...classes)
+    }
+  }
```
Comment (bion) [inline frontend/src/components/animation/TailwindTransition.tsx]: suggestion (blocking): `removeClasses` and `addClasses` don't access anything in the scope of the enclosing function so recommend declaring them outside of it.

File: frontend/src/components/model/BlankModal.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            <path
+              strokeLinecap="round"
+              strokeLinejoin="round"
+              strokeWidth={2}
+              d="M6 18L18 6M6 6l12 12"
+            />
+          </svg>
+        </button>
+      </div>
+      <div className="bg-white m-2 rounded-lg shadow-xl">{children}</div>
```
Comment (bion) [inline frontend/src/components/model/BlankModal.tsx]: suggestion (non-blocking): my read is that this component isn't so much blank as it is a wrapper that puts its children in a modal. consider renaming to something like `ModalWrapper` or even just `Modal`

Comment (bion) [reply frontend/src/components/model/BlankModal.tsx]: Ok woops just saw the next file

Comment (bion) [reply frontend/src/components/model/BlankModal.tsx]: What's this used for?

**Review by bion — APPROVED** [review]

File: frontend/src/components/model/BlankModal.tsx (hunk 4 above)
Comment (deammer) [reply frontend/src/components/model/BlankModal.tsx]: Good point

Comment (deammer) [reply frontend/src/components/model/BlankModal.tsx]: Renamed `Modal` -> `ModalBase` and `BlankModal` -> `PlainModal`

**Pull request merged**